service/matching/matchingEngine.go

```diff
@@ -556,6 +562,81 @@ func (e *matchingEngineImpl) DescribeTaskList(ctx context.Context, request *m.De
 	return tlMgr.DescribeTaskList(request.DescRequest.GetIncludeTaskListStatus()), nil
 }
 
+func (e *matchingEngineImpl) ListTaskListPartitions(ctx context.Context, request *m.ListTaskListPartitionsRequest) (*workflow.ListTaskListPartitionsResponse, error) {
+	aTPartitionInfo, err := e.listTaskListPartitions(request, persistence.TaskListTypeActivity)
+	if err != nil {
+		return nil, err
+	}
+	dTPartitionInfo, err := e.listTaskListPartitions(request, persistence.TaskListTypeDecision)
+	if err != nil {
+		return nil, err
+	}
+	resp := workflow.ListTaskListPartitionsResponse{
+		ActivityTaskPartitions: aTPartitionInfo,
+		DecisionTaskPartitions: dTPartitionInfo,
+	}
+	return &resp, nil
+}
+
+func (e *matchingEngineImpl) listTaskListPartitions(request *m.ListTaskListPartitionsRequest, taskListType int) (*workflow.Partitions, error) {
+	partitionInfo := workflow.Partitions{}
+	partitions, err := e.getAllPartitions(
+		request.GetDomain(),
+		*request.TaskList,
+		taskListType,
+	)
+	if err != nil {
+		return nil, err
+	}
+	partitionHostInfo := make(map[workflow.PartitionKey]workflow.HostInfo, len(partitions))
+
+	if err != nil {
+		return nil, err
+	}
+	for _, partition := range partitions {
+		if host, err := e.getHostInfo(partition); err != nil {
+			partitionInfo := workflow.PartitionKey{PartitionKey: common.StringPtr(partition)}
+			partitionHostInfo[partitionInfo] = host
+		}
+	}
+	return &partitionInfo, nil
+}
+
+func (e *matchingEngineImpl) getHostInfo(partitionKey string) (workflow.HostInfo, error) {
+	keyResolver, err := e.monitor.GetResolver(common.MatchingServiceName)
+	host, err := keyResolver.Lookup(partitionKey)
+	if err != nil {
+		return workflow.HostInfo{}, err
+	}
+	return workflow.HostInfo{HostAddress: common.StringPtr(host.GetAddress())}, nil
+}
+
+func (e *matchingEngineImpl) getAllPartitions(
+	domain string,
+	taskList workflow.TaskList,
+	taskListType int,
+) ([]string, error) {
+	var partitionKeys []string
+	taskListName := taskList.GetName()
+	if strings.HasPrefix(taskListName, taskListPartitionPrefix) {
```

Comment:
see newTaskListName(taskList.GetName()) in tasklist.go - it has an IsRoot() method. You can also add a GetRoot() or Root() method to return the name of the root partition.

Reply:
Added GetRoot method to get the root partition.